karabo_data/reader.py

```diff
@@ -773,16 +777,27 @@ def select_trains(self, train_range):
         ValueError
             If given train IDs do not overlap with the trains in this data.
         """
-        if isinstance(train_range, by_id):
-            start_ix = _tid_to_slice_ix(train_range.value.start, self.train_ids, stop=False)
-            stop_ix = _tid_to_slice_ix(train_range.value.stop, self.train_ids, stop=True)
-            ix_slice = slice(start_ix, stop_ix, train_range.value.step)
-        elif isinstance(train_range, by_index):
-            ix_slice = train_range.value
+        tr = train_range
+        if isinstance(tr, by_id) and isinstance(tr.value, slice):
+            # Slice by train IDs
+            start_ix = _tid_to_slice_ix(tr.value.start, self.train_ids, stop=False)
+            stop_ix = _tid_to_slice_ix(tr.value.stop, self.train_ids, stop=True)
+            new_train_ids = self.train_ids[start_ix:stop_ix:tr.value.step]
+        elif isinstance(tr, by_index) and isinstance(tr.value, slice):
+            # Slice by indexes in this collection
+            new_train_ids = self.train_ids[tr.value]
+        elif isinstance(tr, by_id) and isinstance(tr.value, (list, np.ndarray)):
```

Comment:
Can we allow any iterable? `instance(tr.value, collections.Iterable)`

Reply:
I'm reluctant to accept any iterable because so many things (e.g. strings) are iterable, even when using them wouldn't make sense. Or you could construct e.g. `range(0, int(1e10))`, which will work, but with terrible performance.

Maybe it would make sense to check what types numpy accepts for indexing arrays.

Reply:
> NumPy arrays may be indexed with other arrays (or any other sequence- like object that can be converted to an array, such as lists, with the exception of tuples; see the end of this document for why this is).

So maybe we should be more flexible about this.

A sequence is an ordered collection; karabo_data will currently ignore the order of the indexes given and keep the trains in sequential order. So it may make sense to index by an unordered collection like a set.

Reply:
Out of pragmatism, I think we should leave it to lists and arrays for now: those are the most likely types for people to pass, it's easy to convert another sequence to them, and it avoids any complications with unexpected types matching a broader check.

We can always broaden it later if the limitation turns out to be an inconvenience, but I'd rather make that a practical decision (what do people want to use?) rather than an abstract one.